### PR TITLE
Dynamic ATL

### PR DESCRIPTION
**Pull request by toaster**

### CODE REVIEW

File: spec/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+examples.state
```
Comment (phorsuedzie) [inline spec/.gitignore]: Nicht `/examples.state` ?

Comment (toaster) [inline spec/.gitignore]: Wäre wohl präziser, ja.

**Comment by toaster** [comment]
REVIEW (612b5fd)

File: spec/decision_worker_spec.rb [hunk 2]
```diff
@@ -2,15 +2,23 @@
...
-  let(:atl_config) { { "test" => "atl" } }
-  let(:dtl_config) { "dtl" }
-  let(:unit) { "testt" }
-  let(:default_config) { { unit: unit, decision_task_list: dtl_config, activity_task_lists: atl_config } }
+  let(:default_config) do
+    {
+      unit: "default-unit",
+      decision_task_list: "default-dtl",
+      activity_task_lists: {
+        "default-unit" => "default-unit-atl",
```
Comment (phorsuedzie) [inline spec/decision_worker_spec.rb]: To make this mapping more clear for the spec reader far below, the task list string should better not be too close to the unit name. e.g. `du-atl` or `activity-tasklist-mapped-for-unit-default` would help the lookup of the magic value.

(Generally spoken: examples using a single representant from the wide area of possible values lack some insight about which other (input) representants would be also valid and what's the rule-behind-the-scenes  how input and expectation are connected)

File: spec/decision_worker_spec.rb [hunk 3]
```diff
@@ -2,15 +2,23 @@
...
+        "other-unit" => "other-unit-atl",
+        "test" => "unit_by_workflow_type-atl",
```
Comment (phorsuedzie) [inline spec/decision_worker_spec.rb]: `test`could be "backend" (to express what a unit corresponds in reality) or "gepardenforelle" (to express that it's an arbitrary string).

File: lib/ntswf/decision_worker.rb [hunk 4]
```diff
@@ -99,7 +99,7 @@ def schedule(task, data_providing_event)
       raise Errors::InvalidArgument.new(
           "Missing activity task list config for #{app_in_charge.inspect}") unless task_list
 
-      if task_list_suffix = options['activity_group']
+      if task_list_suffix = options['activity_group'] || activity_group
```
Comment (phorsuedzie) [inline lib/ntswf/decision_worker.rb]: I'm not fully convinced about that.

We have:
- an execution that requests to be run only for a certain activity group
- a decider that maps every execution to a certain activity group
- nothing to have the decider not map to a certain activity group

For the old style execution behaviour, the decider must have no `activity_group` configured. But this is more than we need: we just want some executions to be assigned to an activity group extended task list.

Proposal:
- execution can demand a certain activity group
- execution can demand that decider adds an activity group on its own
- execution does not demand activity group (old behavior).

Code:

```
task-list-suffix =
   case options['activity_group']
   when true
      "-#{activity_group}" # or should the config provide the `-` by itself?
   when nil
      ""
   else
     "-#{options['activity_group']}"
   end

**Comment by phorsuedzie** [comment]
Currently, a worker polls a single activity task list. How difficult would it be to have the worker poll one out of the list configured activity task lists (randomly or round-robin ...). This way, we don't need to wait for the executions assigned to the old stack to terminate. The new stack will just feel responsible for the old executions too, when reconfigured to do so at end-of-lifetime of the old stack.

This way, we don't have to wait for every execution "bound to the old stack" to terminate. As now, we could just shoot the chieftain, because the neighbour tribe will provide the indians to complete open executions.

There will be a a short time were new stack's worker will poll uselessly an empty activity task list, until we drop the old stacks activity task list because we know that there is no open execution left. Even if external executions were not queued with `activity_group: true`, they would just be executed, because there is always a worker listening to every "active" activity task list. But `true` is far better, because the execution will just follow the decider, and not be stuck endlessly with its initial (`true` is the old "always `|| activity_group`" - which also makes an execution follow the decider, but you cannot have old style "no activity group" behaviour then anymore).

File: spec/decision_worker_spec.rb (hunk 3 above)
Comment (toaster) [inline spec/decision_worker_spec.rb]: Yes, I just decided to stay with `test` here because it comes from the workflow execution's name and I wanted to do a stop before rewriting the _whole_ test.

**Comment by toaster** [comment]
Regarding your thoughts on polling multiple activity task lists and dynamically change them during execution:

First, you're not wrong (not completely ;)).
But: It would be far more complexity as it is now, including more manual steps during our future deployment. All for the winning of a few minutes (which will be lost due to the additional manual steps).
I don't think that this is necessary. Maybe the future will proof me wrong, then we could easily add this.

For dynamically change the activity group (as `activity_group: true` would not be real pinning): That's exactly _not_ what we want. I cannot see how you realize two parallel stacks in our planned environment without breaking the seperation of task executions between them with your approach.

File: lib/ntswf/decision_worker.rb (hunk 4 above)
Comment (toaster) [inline lib/ntswf/decision_worker.rb]: YAGNI

Currently we want pinned tasks in one scenario and unpinned (old behaviour) in all others. We don't have a mixed scenario and therefore no need to implement the support for it (and the tests for this support).

Comment (phorsuedzie) [inline lib/ntswf/decision_worker.rb]: The old style executions (e.g. dashboard's own ones) are all *not pinned*, and couldn't be decided by an activity_group tainted (new) decider anymore.

With ( nil | true | <activity_group>), we can still decide every execution.

**Comment by phorsuedzie** [comment]
It's not about saving same minutes. It's about to be able to have all executions (of a unit) to be executed by a single tribe of indians. There would be another approach: If not all indians were equal, but responsible for a dedicated activity_group each, we could just add or remove some activity workers for the activity group on the same stack.

**Comment by phorsuedzie** [comment]
TL;DR: To have executions be bound to different stacks is temporary and not the default case. The default case is, that a single stack is able to exexute everything assigned to it's unit. This should be true even if there are executions that were started with (temporary) stack binding.

**Comment by toaster** [comment]
The use case we need this for is, that an app can run in two different states where no interfering regarding the executions of tasks is allowed.
If we would allow any activity worker to perform all tasks that were originally scheduled on its decision task list (i.e. allow them to perform tasks of other groups, too) we could not strictly separate the executions.
You stated that the workers would only be allowed to execute other group's tasks, when explicitly configured so. While this ensures the separation, it is a higher amount of administration with an advantage that remains theoretical. As long as we don't need this, I don't think we should implement it.

**Comment by phorsuedzie** [comment]
> If we would allow any activity worker to perform all tasks that were originally scheduled on its decision task list (allow them to perform tasks of other groups, too) we could not strictly separate the executions.

Your argumentation is misleading. What I want is to be able to configure a worker differently:

A) only process executions related to a specified activity_group (of its unit)
B) process any execution of its unit

I know that I cannot have A) and B) **at the same time**. But I never required that.

I just wanted to not be stuck with workers only able to do A) from now on for all time. If there is no "different state", I will not care about whether an execution is assigned to any activity group. It should just be processed, because it belongs to the unit the worker/applications is responsible for.

**Comment by toaster** [comment]
For the record:
A worker cannot process multiple activity task lists (at least not in a sane way).
The “app should be able to process tasks from different activity task lists” scenario has to be implemented using separate activity workers for the different activity task lists.

**Comment by dcsaszar** [comment]
FEEDBACK (399a72e)
 
* [ ] Please restore docs for `decision_task_list`.
* [ ] Docs for the `activity_group` feature is missing. So I cannot tell if this is a good approach/naming at all. From reading the code it seems like `options['task_list']` or `options['task_list_suffix']` would be more straight-forward solutions. I'm also confused about the config option with the same name. From reading I get the impression that this is not needed at all, as the suffix could just be set by the task list names configs. So please add docs to clarify the intended usage. (We can also dicuss in Slack beforehand)
* [ ] `_task_lists.dup` it's not obvious from reading the code why this is needed. We should find a more readable solution. Note: Probably manipulating the config was a bad design from the start.
* [ ] https://github.com/infopark/ntswf/pull/26/files#diff-fe2fd304b2644181e2a3295ab38adeb1R187 is untested, but cannot be deleted because there is code relying on it. Either delete it and fix the then-broken code (my favorite solution), or add a test.
* [ ] Method naming: `activity_group_list` returns an activity task list. It expects an argument `task_list`, which is an activity task list, when called by `activity_task_list`. This is *very* confusing. Find better names.
* [ ] Method args: `activity_group_list` receives a task list name, but this always comes from `activity_task_lists`. So it can be simplified, it only needs the unit name. (If this method still survives the other feedback at all)
* [ ] Overriding `self.activity_group` with a variable of the same name but different content is confusing and should be avoided.

**Comment by toaster** [comment]
REVIEW (af89ff9)

Note: version bump was resorted to the end and force-pushed.

Manipulating an non duplicated config was definitely a bad design.
The refactoring to improve this is out of focus of this PR.

**Comment by toaster** [comment]
REVIEW (91b07f2)

File: lib/ntswf/activity_worker.rb [hunk 5]
```diff
@@ -48,7 +48,8 @@ def process_activity_task
     protected
 
     def activity_task_list
-      activity_task_lists[default_unit] or raise Errors::InvalidArgument.new(
+      @activity_task_list ||= activity_task_lists[default_unit] or
```
Comment (dcsaszar) [inline lib/ntswf/activity_worker.rb]: Do we really want the memoizing here? It appears this only makes the code more complex without a gain.

File: lib/ntswf/base.rb [hunk 6]
```diff
@@ -77,9 +81,12 @@ def activity_task_lists
...
     end
 
+    def decision_task_lists
+      @config.decision_task_lists
+    end
+
     def decision_task_list
-      @config.decision_task_list or raise Errors::InvalidArgument.new(
-          "Missing decision task list configuration")
+      decision_task_lists[default_unit]
```
Comment (dcsaszar) [inline lib/ntswf/base.rb]: Why did we remove the "fail on missing config"?

File: lib/ntswf/client.rb [hunk 7]
```diff
@@ -10,6 +10,11 @@ module Client
     # The options configure the control flow of the task.
     # Excluding *:execution_id* they will be stored in the *input* argument of the task as JSON.
     # @param options [Hash] The task's options. Keys with special meaning:
+    # @option options [String] :activity_task_list
+    #   The activity task list that activity tasks should belong to. May be +nil+.
+    #   There must be workers configured with the same activity task list to handle the activity
+    #   tasks.
+    #   See {Base#configure}'s :activity_task_lists.
```
Comment (dcsaszar) [inline lib/ntswf/client.rb]: * We should add a note about having precedence over `:unit`.
* We could drop the note "There must be workers configured with the same activity task list to handle the activity tasks. See x.". This is obvious, and not related to this particular option.
* "activity tasks should belong to" => "this activity task belongs to"

File: lib/ntswf/client.rb [hunk 8]
```diff
@@ -72,7 +77,7 @@ def start_swf_workflow_execution(options)
         execution_start_to_close_timeout: 48 * 3600,
         input: options.to_json,
         tag_list: [options[:unit].to_s, options[:name].to_s] + Array(options[:tag_list]),
-        task_list: decision_task_list,
+        task_list: decision_task_lists[options[:unit]] || decision_task_list,
```
Comment (dcsaszar) [inline lib/ntswf/client.rb]: Can we refactor this into the `decision_task_list` method?

File: lib/ntswf/decision_worker.rb [hunk 9]
```diff
@@ -94,10 +94,11 @@ def retry_or_continue_as_new(task, event)
     def schedule(task, data_providing_event)
       input = data_providing_event.attributes.input
       options = parse_input(input)
-      app_in_charge = options['unit'] || guess_app_from(data_providing_event)
-      task_list = activity_task_lists[app_in_charge]
-      raise Errors::InvalidArgument.new(
-          "Missing activity task list config for #{app_in_charge.inspect}") unless task_list
+      task_list = options['activity_task_list'] || (
```
Comment (dcsaszar) [inline lib/ntswf/decision_worker.rb]: Please make the part in braces a (or several) methods. This will decrease code complexity.

File: lib/ntswf/activity_worker.rb (hunk 5 above)
Comment (toaster) [inline lib/ntswf/activity_worker.rb]: I just reverted (AFAIR) but will remove the memoizing.

File: lib/ntswf/base.rb (hunk 6 above)
Comment (toaster) [inline lib/ntswf/base.rb]: Ooops … will reintroduce it

File: lib/ntswf/client.rb (hunk 8 above)
Comment (toaster) [inline lib/ntswf/client.rb]: will have a look at it

File: lib/ntswf/decision_worker.rb (hunk 9 above)
Comment (toaster) [inline lib/ntswf/decision_worker.rb]: No, it won't. It just adds one or more methods without a clear mission.
The only meaningful method I currently can think of would be `activity_task_list_for_event(event)` which would include the the lines above, too, and duplicate the input extraction.
Should I go on in this direction?

**Comment by toaster** [comment]
REVIEW (5455081)

File: lib/ntswf/decision_worker.rb (hunk 9 above)
Comment (dcsaszar) [inline lib/ntswf/decision_worker.rb]: Yes please.
I'm thinking especially in reducing complexity by reducing variables. E. g. 
```ruby
def task_list
  enforced_task_list || unit_task_list || guessed_task_list
end
```

File: lib/ntswf/client.rb [hunk 10]
```diff
@@ -11,10 +11,9 @@ module Client
     # Excluding *:execution_id* they will be stored in the *input* argument of the task as JSON.
     # @param options [Hash] The task's options. Keys with special meaning:
     # @option options [String] :activity_task_list
-    #   The activity task list that activity tasks should belong to. May be +nil+.
-    #   There must be workers configured with the same activity task list to handle the activity
-    #   tasks.
-    #   See {Base#configure}'s :activity_task_lists.
+    #   The activity task list that this execution's activity tasks should belong to. May be +nil+.
+    #   If not given the decider will use the activity task list configured for the unit of this
+    #   execution.
```
Comment (dcsaszar) [inline lib/ntswf/client.rb]: * > The activity task list that this execution's activity tasks should belong to.

  Most of the time it's only a single execution, so we should not confuse the casual reader. How about simplifying to sth like

  `The activity task list this task is assigned to.`?

* I'm also not sure if we should go into the details of the decider when documenting the client. The idea of the separation was to make it easy to use, without having to know the inner workings of SWF. 
  `If not given, see :unit`?

* We must also add docs on how the `:unit` option behaves regarding the new `decision_task_lists`.

Comment (dcsaszar) [inline lib/ntswf/client.rb]: Issues still open:

* [ ] Further improve docs
* [ ] improve code readability: try to get rid of `dup`
* [ ] reduce code complexity/improve readability , see https://github.com/infopark/ntswf/pull/26#discussion_r79637216

Comment (toaster) [inline lib/ntswf/client.rb]: * I did not write anything about multiple executions. Also one doesn't start a task but an execution. Therefore my wording is right.
* NTSWF is not easy to use … at least it is not usable without understanding SWF (you have to setup and configure it).

Comment (toaster) [inline lib/ntswf/client.rb]: I won't mess around with the configuration thing. It is messed up with this `OpenStruct` and I only corrected a misbehavior (changing input hash).

**Comment by toaster** [comment]
REVIEW (77ee9e9)

The activity task list computation was outsourced a little bit. I think it is sufficient to let it as it is now.

File: lib/ntswf/client.rb (hunk 10 above)
Comment (dcsaszar) [inline lib/ntswf/client.rb]: > The activity task list that this execution's activity tasks should belong to.

task*s* indicates multiple tasks. But it's only a single task, except for the rare case of recurring executions.

File: lib/ntswf/base.rb [hunk 11]
```diff
@@ -77,9 +81,21 @@ def activity_task_lists
       @config.activity_task_lists
     end
 
-    def decision_task_list
```
Comment (dcsaszar) [inline lib/ntswf/base.rb]: We should not remove public API in a minor version upgrade

File: lib/ntswf/base.rb [hunk 12]
```diff
@@ -77,9 +81,21 @@ def activity_task_lists
...
-      @config.decision_task_list or raise Errors::InvalidArgument.new(
-          "Missing decision task list configuration")
+    def activity_task_list(unit: nil)
```
Comment (dcsaszar) [inline lib/ntswf/base.rb]: Not sure if we really want to make this public API. Either:
* move it to protected, or
* make the interface more self-explanatory, like `activity_task_list_for_unit`

It's also breaking existing API.

File: lib/ntswf/client.rb [hunk 13]
```diff
@@ -10,6 +10,9 @@ module Client
     # The options configure the control flow of the task.
     # Excluding *:execution_id* they will be stored in the *input* argument of the task as JSON.
     # @param options [Hash] The task's options. Keys with special meaning:
+    # @option options [String] :activity_task_list
+    #   The activity task list that this execution's activity tasks should belong to.
```
Comment (dcsaszar) [inline lib/ntswf/client.rb]: Improve wording.
* activity task (not tasks)
* no "should", it's mandatory
* "routed to" (official SWF wording) instead of "belong to"

File: lib/ntswf/base.rb [hunk 14]
```diff
@@ -145,8 +162,17 @@ def raise_if_invalid_task_list
...
+        @config.decision_task_lists = @config.decision_task_lists.dup
+        @config.decision_task_lists.each do |key, value|
+          @config.decision_task_lists[key] = autocomplete(value, "#{key}-dtl")
+        end
+      else
+        @config.decision_task_lists =
+            {default_unit => autocomplete(@config.decision_task_list, "master-dtl")}
+      end
       if @config.activity_task_lists
+        @config.activity_task_lists = @config.activity_task_lists.dup
```
Comment (dcsaszar) [inline lib/ntswf/base.rb]: Still unclear why we need the `.dup`s here. I'd really like to make this code more self-explanatory.

File: lib/ntswf/base.rb (hunk 12 above)
Comment (toaster) [inline lib/ntswf/base.rb]: The backend needs it, therefore it has to be public API. That's why I explicitly tested it.
Also it is not more self-explanatory when adding `_for_unit` since this is misleading in the default case.
Last but not least it is analogue to `decision_task_list` which is public API, too, and was by no means deleted.

File: lib/ntswf/client.rb (hunk 13 above)
Comment (toaster) [inline lib/ntswf/client.rb]: A single activity task is a special case while “tasks” is correct and implies this special case.
Again: NTSWF is of no use if one has not understood what SWF is and how it works in general.

File: lib/ntswf/base.rb (hunk 11 above)
Comment (toaster) [inline lib/ntswf/base.rb]: Not removed just resorted and extended (for internal use).

File: lib/ntswf/base.rb (hunk 14 above)
Comment (toaster) [inline lib/ntswf/base.rb]: Don't hesitate to do so … in a separate branch. This `dup` is necessary because we would change the input hash otherwise.
In production this may not be a problem. In the (new) tests it is.

**Comment by toaster** [comment]
REVIEW (08a2d3c)
